Simplify and document expansion functions of Graph

patowen opened the pull request:
This PR accomplishes a few things, all in the spirit of hopefully reducing the mental load of understanding all the algorithms that allow `Graph` to grow:
- The term "parent" no longer depends on the graph's generation order. It now refers to the node along the first descender in enum order.
- Uses of the term "parent" and "child" were removed whenever they didn't refer to the above meaning of "parent". This includes renaming `insert_child` to `insert_neighbor` and renaming `populate_shorter_neighbors_of_child` to `populate_shorter_neighbors_of_subject` (where "subject" is a term that was introduced internally for "the neighbor node we want to insert" to try to improve clarity).
- An unused code path was removed in `ensure_neighbor` to cover the impossible scenario of a not-yet-generated shorter neighbor. Removing this code-path turned `ensure_neighbor` into a one-liner.
- The most conceptually difficult function in the algorithm, `populate_shorter_neighbors_of_subject` is now documented in detail. For simplification purposes, it was also altered to return all shorter neighbors, including the passed in argument (as it previously excluded this argument, resulting in extra logic for `insert_neighbor`).

patowen commented:
I believe some of the comments I've written don't make sense, although it's hard to tell in what way they don't make sense and how to improve them, so I hope that peer review will help here.

Ralith commented on common/src/graph.rs:
```diff
@@ -159,62 +159,39 @@ impl Graph {
+
+        // Select the side along the canonical path from the origin to this node. This is guaranteed
+        // to be the first entry of the `shorter_neighbors_of_neighbor` iterator.
+        let (parent_side, parent) = shorter_neighbors_of_neighbor.clone().next().unwrap();
```
```suggestion
        let (parent_side, parent) = shorter_neighbors_of_neighbor.next().unwrap();
```

Ralith commented on common/src/graph.rs:
```diff
@@ -231,27 +208,49 @@ impl Graph {
-            count += 1;
         }
-        (0..2).filter_map(move |i| neighbors[i])
+        (0..3).filter_map(move |i| shorter_neighbors_of_subject[i])
```
While we're revisiting this, might as well take advantage of the `[T]::into_iter` that's been introduced since it was written, just to be a bit less weird:
```suggestion
        shorter_neighbors_of_subject.into_iter().filter_map(|x| x)
```

patowen replied:
Thanks! That does seem nice. Clippy further simplifies `filter_map(|x| x)` to `flatten()`, which seems a bit weird to me, but that's only because I consider the concept of an `Option` being a type of iterator to be a bit weird conceptually. I'll trust the designers of Rust that this is okay.

Ralith replied:
Think of it as a container of 0 or 1 items, just like an array is a container of N items, or a Vec is a container of an arbitrary number of items.

patowen replied:
Oh, yeah. I do think of it that way. It's just not the most natural way for me.

Ralith commented on common/src/graph.rs:
```diff
@@ -231,27 +208,49 @@ impl Graph {
-        parent_side: Side,
+        node: NodeId,
+        side: Side,
     ) -> impl Iterator<Item = (Side, NodeId)> + Clone {
```
Do we still need `+ Clone` on this?

Ralith replied:
On second thought, do we need to return an iterator at *all*? It seems like we only care about the parent, so this could probably be simplified to just return `(Side, NodeId)` directly.

patowen replied:
We care about all elements of this iterator because of the
```rust
for (side, neighbor) in shorter_neighbors_of_neighbor
```
loop near the bottom of `insert_neighbor`. This is also why the iterator needs to be cloned, because it's first used to find the "parent", and then it is used to perform all the linking of neighbors.

Ralith commented on common/src/graph.rs:
```diff
@@ -159,62 +159,39 @@ impl Graph {
+        let length = self.nodes[&node].length + 1;
         self.nodes
-            .insert(id, NodeContainer::new(Some(side), length));
-        self.link_neighbors(id, parent, side);
```
Why was this not necessary?

Ralith replied:
Oh, I see.

Ralith commented on common/src/graph.rs:
```diff
@@ -159,62 +159,39 @@ impl Graph {
+        // An assumption made by this function is that `side` is not a descender of `node`. This is a safe assumption to make
+        // because a node cannot be constructed before its shorter neighbors. The call to `populate_shorter_neighbors_of_neighbor`
+        // guarantees this.
+        let shorter_neighbors_of_neighbor = self.populate_shorter_neighbors_of_neighbor(node, side);
```
I found this variable name confusing. I understood it to mean "shorter neighbors of the neighbor [of the new node]", but in fact it seems to be "shorter neighbors [of the new node]". Maybe standardize on `subject` like below?

patowen replied:
I'm not surprised it's confusing. It's been a while, so looking at this PR again with fresh eyes should hopefully help me think of better variable names here.

patowen replied:
Thinking on this again, I agree with you that standardizing on the name `subject` makes sense, as long as that naming decision doesn't leak outside the `graph` module. I'll go ahead with your suggestion.

Ralith commented:
Great improvement overall! I don't think I ever actually understood this logic, so this clarity is striking.

patowen commented:
Ah, the "Compare" button isn't as useful this time because it also includes all the changes from `master` that were included. I'll list out the changes I made:

- At the top of `insert_neighbor`, I added the comment
```rust
// To help improve readability, we use the term "subject" to refer to the not-yet-created neighbor node, since the term.
// "neighbor" is already used in many other places.
```
- In `insert_neighbor, I renamed `shorter_neighbors_of_neighbor` to `shorter_neighbors_of_subject`.
- I changed the name of the function `populate_shorter_neighbors_of_neighbor` with `populate_shorter_neighbors_of_subject` and updated the function header comment to explain what "subject" means.
- I removed the comment at the top of the `populate_shorter_neighbors_of_subject` function explaining the use of "subject", as the concept is no longer introduced there.
- I added the following addendum to `populate_shorter_neighbors_of_subject`'s long comment:
```rust
// Note that this shorter neighbor might not yet be in the graph, which is why we call `ensure_neighbor`
// here instead of `neighbor`. This means that nodes in the graph will be recursively created as necessary.
```
- I used `shorter_neighbors_of_subject.into_iter().flatten()` as suggested.

patowen commented:
I realize that my note about calling `ensure_neighbor` instead of `neighbor` was unnecessarily verbose for what is already a long and dense explanation, so I replaced it with the `(and recursively create if needed)` parenthetical, which accomplishes the same thing in much fewer words.
